File: src/codegen.go
package main

import (
	"strings"
)

// GenBash compiles the sequence of parsed Tokens into a valid, executable Bash script string.
func GenBash(parsedItems []Token) string {
	out := []string{BashHeaders}
	var blockStack []string
	var deferStack []string

	for _, item := range parsedItems {
		kind, val := item.Type, item.Value

		switch kind {
		case "RAW_CMD", "EMPTY", "ASSIGN", "LOCAL_ASSIGN", "CONST", "DEL", "PRINTN", "PRINT", "RAW", "MATH":
			out = append(out, val)

		case "IF":
			cond := processCond(val)
			if strings.TrimSpace(cond) == "" {
				cond = "true"
			}
			out = append(out, "if [[ "+cond+" ]]; then")
			blockStack = append(blockStack, "if")

		case "ELIF":
			cond := processCond(val)
			if strings.TrimSpace(cond) == "" {
				cond = "true"
			}
			out = append(out, "elif [[ "+cond+" ]]; then")

		case "ELSE":
			out = append(out, "else")

		case "WHILE":
			out = append(out, "while [[ "+val+" ]]; do")
			blockStack = append(blockStack, "loop")

		case "UNTIL":
			out = append(out, "until [[ "+val+" ]]; do")
			blockStack = append(blockStack, "loop")

		// Handle for loop block headers
		case "FOR":
			out = append(out, "for "+val+"; do")
			blockStack = append(blockStack, "loop")

		case "DEFER":
			deferStack = append([]string{val}, deferStack...)

		case "CASE_START":
			if strings.HasPrefix(val, "\"$") {
				out = append(out, "case "+val+" in")
			} else {
				out = append(out, "case \"$"+val+"\" in")
			}
			blockStack = append(blockStack, "case")

		case "CASE_OVER":
			out = append(out, ";;")

		case "FN_START":
			out = append(out, val)
			blockStack = append(blockStack, "fn")

		case "RETURN":
			out = append(out, "    "+val)

		case "BLOCK_END":
			if len(blockStack) == 0 {
				out = append(out, "}")
				continue
			}

			lastBlock := blockStack[len(blockStack)-1]
			blockStack = blockStack[:len(blockStack)-1]

			switch lastBlock {
			case "if":
				out = append(out, "fi")
			case "loop":
				out = append(out, "done")
			case "case":
				out = append(out, "esac")
			case "fn":
				out = append(out, "}")
			}
		}
	}

	// Prepend deferred tasks as an EXIT trap if any were registered
	if len(deferStack) > 0 {
		deferString := strings.Join(deferStack, "; ")
		trapCmd := "\ntrap '" + deferString + "' EXIT\n"

		out = append(out[:1], append([]string{trapCmd}, out[1:]...)...)
	}

	return strings.Join(out, "\n")
}

// processCond normalizes condition expressions and replaces logical words with Bash operators.
func processCond(val string) string {
	val = strings.TrimSpace(val)
	val = strings.ReplaceAll(val, " and ", " && ")
	val = strings.ReplaceAll(val, " or ", " || ")

	words := strings.Fields(val)
	processed := make([]string, 0, len(words))

	for _, w := range words {
		if isIdentifier(w) && w != "true" && w != "false" {
			processed = append(processed, "\"$"+w+"\"")
		} else {
			processed = append(processed, w)
		}
	}

	return strings.Join(processed, " ")
}

File: src/codegen.py
# codegen.py
from config import BASH_HEADERS

def gen_bash(parsed_items):
    out = [BASH_HEADERS]
    block_stack = []
    defer_stack = []

    for kind, val in parsed_items:
        if kind in ("RAW_CMD", "EMPTY", "ASSIGN", "LOCAL_ASSIGN", "CONST", "DEL", "PRINTN", "PRINT", "RAW"):
            out.append(val)

        elif kind == "IF":
            cond = " ".join([f'"${w}"' if w.isidentifier() and w not in ("true", "false") else w for w in val.strip().replace(" and ", " && ").replace(" or ", " || ").split()])
            out.append(f"if [[ {cond} ]]; then")
            block_stack.append("if")

        elif kind == "ELIF":
            cond = " ".join([f'"${w}"' if w.isidentifier() and w not in ("true", "false") else w for w in val.strip().replace(" and ", " && ").replace(" or ", " || ").split()])
            out.append(f"elif [[ {cond} ]]; then")

        elif kind == "ELSE":
            out.append("else")

        elif kind == "WHILE":
            out.append(f"while [[ {val} ]]; do")
            block_stack.append("loop")

        elif kind == "UNTIL":
            out.append(f"until [[ {val} ]]; then")
            block_stack.append("loop")

        elif kind == "DEFER":
            defer_stack.insert(0, val)
            continue

        elif kind == "CASE_START":
            out.append(f"case \"${val}\" in")
            block_stack.append("case")

        elif kind == "CASE_OVER":
            out.append(";;")

        elif kind == "FN_START":
            out.append(val)

        elif kind == "RETURN":
            out.append(f"   {val}")

        elif kind == "BLOCK_END":
            if not block_stack:
                out.append("}")  # Обычная закрывающая скобка (например, для функций)
                continue

            last_block = block_stack.pop()
            if last_block == "if":
                out.append("fi")
            elif last_block == "loop":
                out.append("done")
            elif last_block == "case":
                out.append("esac")

    if defer_stack:
        defer_string = "; ".join(defer_stack)
        out.insert(1, f"\ntrap '{defer_string}' EXIT\n")

    return "\n".join(out)
